### PR TITLE
chain-storage-sqlite: implement in-memory index

### PR DESCRIPTION
**Pull request by eugene-babichenko**

### CODE REVIEW

**Comment by eugene-babichenko** [comment]
Working on some benchmarks to see how does it actually perform

File: chain-storage-sqlite/Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,7 @@ chain-core = { path = "../chain-core" }
 chain-storage = { path = "../chain-storage" }
 r2d2 = { version = "0.8" }
 r2d2_sqlite = { version = "0.8" }
+thiserror = "1.0"
```
[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply chain-storage-sqlite/Cargo.toml]: Well, I've seen a couple of different libraries used throughout the project for doing it (I believe that `strum`is the third one I encountered) so I just got confused a bit.

File: chain-storage-sqlite/src/lib.rs [hunk 2]
```diff
@@ -30,23 +131,21 @@ where
                   begin;
 
                   create table if not exists BlockInfo (
-                    hash blob primary key,
```
[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply chain-storage-sqlite/src/lib.rs]: Well, I guess that not really. Writes are not so frequent that we are severely affected by the sqlite index. But on the other hand do we really need that given that we have the in-memory index now?

Possibly a good idea is having this as an option in the future. I mean switching between sqlite and in-memory indexes.

File: chain-storage-sqlite/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (vincenthz) [reply chain-storage-sqlite/Cargo.toml]: Ok, we'll need to sort this out eventually. I'm completely allergic to dependencies :)

File: chain-storage-sqlite/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vincenthz) [reply chain-storage-sqlite/src/lib.rs]: Ideally yes, once we have mature APIs in place, we'll be able to abstract the backend properly and have runtime switches. At least that was the dream with chain-storage. At the moment we only need one backend that works well enough, so we don't need the APIs in place right away.

**Pull request merged**